### ISSUE TITLE
CygWin local paths not found

### ISSUE THREAD

**Issue by cesarjorgemartinez** [hashicorp/packer#5559]
CygWin local paths not found

I'm using CygWin (64) and when use local ISO's then these files not founded.

```
$ ls -l /cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso
-rwxrwx---+ 1 admin None 830472192 sep 18 09:19 /cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso
```

- Packer version from 1.1.1
- Host platform
Windows 7 and CygWin
- Debug log output from `PACKER_LOG=1 packer build template.json`.
```
...
    virtualbox-iso: Downloading or copying: file:///cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso
    virtualbox-iso: Error downloading: CreateFile cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso: The system can not find the path specified.
...
```
- The _simplest example template and scripts_ needed to reproduce the bug.
```
{
  "variables": {
    "guest_os_type": "RedHat_64",
    "iso": "file:///cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso",
    "checksum": "5848f2fd31c7acf3811ad88eaca6f4aa"
  },
  "builders": [
    {
      "type": "virtualbox-iso",
      "guest_os_type": "{{ user `guest_os_type` }}",
      "iso_url": "{{ user `iso` }}",
      "iso_checksum": "{{ user `checksum` }}",
      "iso_checksum_type": "md5",
      "vm_name": "MyCentOS7",
      "ssh_username": "user",
      "ssh_password": "pass",
      "ssh_wait_timeout": "20m",
      "http_directory": "http",
      "boot_command": [
        "<tab> text ks=http://{{ .HTTPIP }}:{{ .HTTPPort }}/anaconda-ks.cfg<enter><wait>"
      ],
      "shutdown_command": "echo 'packer' | sudo -S shutdown -P now"
    }
  ]
}
```

**Comment by mwhooker** [comment]
please rerun with `PACKER_LOG=1` and post the full debug output

**Comment by cesarjorgemartinez** [comment]
[fulldebugpackercygwin.txt](https://github.com/hashicorp/packer/files/1453092/fulldebugpackercygwin.txt)

**Comment by cesarjorgemartinez** [comment]
An observation about this variable: I need to do: export PACKER_LOG=1

**Comment by mwhooker** [comment]
wondering if this will be solved by #5570, but since I don't see the error off-hand, I can't say for sure

**Comment by mwhooker** [comment]
@cesarjorgemartinez could you try adding `"target_path": "C:/cygdrive/c/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso"` to your builder? I wonder if that might fix it

**Comment by cesarjorgemartinez** [comment]
I do not know how to test this PR 5570, I do not know how to compile that in Windows.

**Comment by cesarjorgemartinez** [comment]
Using file:///C:/software/SO/CENTOS/CentOS-7-x86_64-Minimal-1708.iso (Windows style path), is a workaround.

**Comment by SwampDragons** [comment]
There is a [lot of special casing](https://github.com/hashicorp/packer/blob/79fcd40bfcafad3c3fb45273b0746c983d69d2c6/common/config.go#L69-L82) in this code around windows paths; looks like the parser does not consider "/foo/bar/baz" an absolute path if packer is running on windows.  It explicitly looks for the drive when determining whether something is an absolute path -- which makes sense to me and I don't think it's a bug.  So the parser treats the url you gave it as a relative path, and that relative path doesn't exist -- hence the confusion and frustration.  There may be a better way to handle this, but I don't think the path given is complete enough for packer to handle appropriately.

**Comment by cesarjorgemartinez** [comment]
I am running in Windows more CygWin:
$ uname -a
CYGWIN_NT-6.1 hostname 2.9.0(0.318/5/3) 2017-09-12 10:18 x86_64 Cygwin

**Comment by testworksau** [comment]
I have the same issue with Packer 1.1.2 from a windows 10 host in the regular command prompt.

Version info:

```
2017/11/21 17:04:09 packer.exe: 2017/11/21 17:04:09 [INFO] Packer version: 1.1.2
2017/11/21 17:04:09 packer.exe: 2017/11/21 17:04:09 Packer Target OS/Arch: windows amd64
2017/11/21 17:04:09 packer.exe: 2017/11/21 17:04:09 Built with Go Version: go1.9
```

I set `iso_url` to:

```
"iso_url" : "Windows2016.iso",
```

Directory contents are:

```
C:\bitbucket\packer_ahv>dir
...
 Directory of C:\bitbucket\packer_ahv

21/11/2017  04:39 PM    <DIR>          .
21/11/2017  04:39 PM    <DIR>          ..
21/11/2017  01:06 PM    <DIR>          ahv
21/11/2017  12:01 PM    <DIR>          answer_files
21/11/2017  04:39 PM    <DIR>          packer_cache
21/11/2017  02:12 PM    <DIR>          scripts
16/10/2017  01:21 PM     6,640,936,960 Windows2016.iso
               1 File(s)  6,640,936,960 bytes
               6 Dir(s)  616,058,699,776 bytes free
```

Command output while specifying the valid local file:

```
C:\bitbucket\packer_ahv>c:\tools\packer.exe build ahv/tcwintestagent.json
TCWinTestAgent output will be in this color.

==> TCWinTestAgent: Downloading or copying ISO
    TCWinTestAgent: Downloading or copying: file://C:/bitbucket/packer_ahv/Windows2016.iso
    TCWinTestAgent: Error downloading: CreateFile bitbucket/packer_ahv/Windows2016.iso: The system cannot find the path specified.
==> TCWinTestAgent: ISO download failed.
Build 'TCWinTestAgent' errored: ISO download failed.

==> Some builds didn't complete successfully and had errors:
--> TCWinTestAgent: ISO download failed.

==> Builds finished but no artifacts were created.
```

Note that if you actually enter an invalid filename, the output is different (the CWD is not prepended to the part where it says "Downloading or copying file"). In the example below, I set `iso_url` to:

```
"iso_url" : "zWindows2016.iso",
```

And received:

```
C:\bitbucket\packer_ahv>c:\tools\packer.exe build ahv/tcwintestagent.json
TCWinTestAgent output will be in this color.

==> TCWinTestAgent: Downloading or copying ISO
    TCWinTestAgent: Downloading or copying: file://zWindows2016.iso
    TCWinTestAgent: Error downloading: Stat : The system cannot find the path specified.
==> TCWinTestAgent: ISO download failed.
Build 'TCWinTestAgent' errored: ISO download failed.

==> Some builds didn't complete successfully and had errors:
--> TCWinTestAgent: ISO download failed.

==> Builds finished but no artifacts were created.
```

Using a network fileshare in the works fine.

Here's the pertinent part of the build log when run from bash:

```
2017/11/21 17:10:53 Preparing build: TCWinTestAgent
2017/11/21 17:10:53 packer.exe: 2017/11/21 17:10:53 use specified accelerator: kvm
2017/11/21 17:10:53 Waiting on builds to complete...
2017/11/21 17:10:53 Starting build run: TCWinTestAgent
2017/11/21 17:10:53 Running builder: qemu
2017/11/21 17:10:53 [INFO] (telemetry) Starting builder qemu
2017/11/21 17:10:53 packer.exe: 2017/11/21 17:10:53 Qemu path: C:\Program Files\qemu\qemu-system-x86_64.exe, Qemu Image page: C:\Program Files\qemu\qemu-img.exe
2017/11/21 17:10:53 ui: ==> TCWinTestAgent: Downloading or copying ISO==> TCWinTestAgent: Downloading or copying ISO

2017/11/21 17:10:53 ui:     TCWinTestAgent: Downloading or copying: file://C:/bitbucket/packer_ahv/Windows2016.iso    TCWinTestAgent: Downloading or copying: file://C:/bitbucket/packer_ahv/Windows2016.iso
    TCWinTestAgent: Error downloading: CreateFile bitbucket/packer_ahv/Windows2016.iso: The system cannot find the path specified.

2017/11/21 17:10:53 packer.exe: 2017/11/21 17:10:53 Acquiring lock to download: file://C:/bitbucket/packer_ahv/Windows2016.iso
2017/11/21 17:10:53 packer.exe: 2017/11/21 17:10:53 Parsed URL: &url.URL{Scheme:"file", Opaque:"", User:(*url.Userinfo)(nil), Host:"C:", Path:"/bitbucket/packer_ahv/Windows2016.iso", RawPath:"", ForceQuery:false, RawQuery:"", Fragment:""}
2017/11/21 17:10:53 packer.exe: 2017/11/21 17:10:53 [DEBUG] Using local file: /bitbucket/packer_ahv/Windows2016.iso
==> TCWinTestAgent: ISO download failed.2017/11/21 17:10:53 ui:     TCWinTestAgent: Error downloading: CreateFile bitbucket/packer_ahv/Windows2016.iso: The system cannot find the path specified.

2017/11/21 17:10:53 ui error: ==> TCWinTestAgent: ISO download failed.Build 'TCWinTestAgent' errored: ISO download failed.


==> Some builds didn't complete successfully and had errors:
2017/11/21 17:10:53 [INFO] (telemetry) ending qemu
2017/11/21 17:10:53 [INFO] (telemetry) found error: ISO download failed.
2017/11/21 17:10:53 ui error: --> TCWinTestAgent: ISO download failed.
Build 'TCWinTestAgent' errored: ISO download failed.

2017/11/21 17:10:53 Builds completed. Waiting on interrupt barrier...
2017/11/21 17:10:53 machine readable: error-count []string{"1"}
2017/11/21 17:10:53 ui error:
==> Builds finished but no artifacts were created.
==> Some builds didn't complete successfully and had errors:
2017/11/21 17:10:53 machine readable: TCWinTestAgent,error []string{"ISO download failed."}
2017/11/21 17:10:53 ui error: --> TCWinTestAgent: ISO download failed.
```

Using full paths to the file doesn't appear to help either.

**Comment by SwampDragons** [comment]
@testworksau Your problem seems to be related to #5589 rather than this ticket.  It's already been fixed on master; if you want to use the dev binary provided on the other ticket, you can, or else you can downgrade until we can release the next version. Apologies for the regression.

**Comment by SwampDragons** [comment]
@cesarjorgemartinez oh hm I see.  I haven't used CygWin before.  So it looks like the pattern here is that `/cygdrive/c/foo/bar/baz.txt` should map to `C://foo/bar/baz.txt`?  and if it was `/cygdrive/d/blahblah.txt` then it should map to `D://blahblah.txt`

**Comment by SwampDragons** [comment]
@cesarjorgemartinez Can you try out this build and let me know if it solves your pathing issues?
 [packer.zip](https://github.com/hashicorp/packer/files/1493207/packer.zip)

**Comment by SwampDragons** [comment]
Oops -- just updated the binary there because I made a dumb mistake

**Comment by brupelo** [comment]
Little note regarding to this issue, on the recently closed [ticket](https://github.com/hashicorp/packer/issues/5628) I had opened (possible duplicate of this one) I had used conemu to test few versions of packer (the official one from website and the latest master one 4/5 days ago). Good thing about [conemu](https://conemu.github.io/) is that testing different terminals like cygwin/msys/cmd/powershell/vs_cmd_prompt becomes a pleasant experience. 

I'm commenting this as I assume you guys agree that packer should work fine on different terminal flavours, not just unix-style ones like cygwin, right?

**Comment by SwampDragons** [comment]
@brupelo that's a cool site - thanks for the recommendation.

**Comment by SwampDragons** [comment]
@cesarjorgemartinez have you had a chance to try the branch I built for you?

**Comment by SwampDragons** [comment]
Since no one is interested in testing this patch, I am going to close the PR and the issue. I'll reopen if anyone gives feedback.

**Comment by hoshsadiq** [comment]
Hello, this is obviously a lot later. Having the same issue. Will be happy to test your branch, but I think a more recent version might be worth using. By the way, had a quick look at the PR. There's a better way to convert the paths. There's a cygpath utility that comes preinstalled on cygwin. Help output is here: https://cygwin.com/cygwin-ug-net/cygpath.html

**Comment by hoshsadiq** [comment]
Hello, I know this has been closed, but figured I'd reply here first.

Having the same issue. Will be happy to test your branch, but I think a more recent version might be worth using. By the way, had a quick look at the PR. There's a better way to convert the paths. There's a cygpath utility that comes preinstalled on cygwin. Help output is here: https://cygwin.com/cygwin-ug-net/cygpath.html

The gist is:
```
cygpath -w $PATH # /cygdrive/c/bla > C:\bla
cygpath -m $PATH # /cygdrive/c/bla > C:/bla
cygpath -u $PATH # C:\bla > /cygdrive/c/bla
```

**Comment by SwampDragons** [comment]
Thanks @hoshsadiq!  I'll try to get an updated version of that patch for you to test.

**Comment by SwampDragons** [comment]
In the last release, we totally reworked how we do downloads, so that old patch is useless anyway.  

We now use a third party tool to perform our downloads, so this may now count as an upstream bug. But just to be sure -- have you tried using packer v1.4.0 to perform this retrieval?

**Comment by hoshsadiq** [comment]
Yep, the version I'm running is 1.4.0. Also, this is not necessarily about downloading. This also affects e.g. file provisioner which is where I got the error.

**Comment by anshulsharma-hashicorp** [comment]
Hi @cesarjorgemartinez and @hoshsadiq Thanks for reporting this.

This issue is coming because you are trying to add the cygpath for the iso file or any other file may be in provisioner. the packer does not handle the cygpath as of now and there is work around for this is to give the windows standard path, which packer can work with.
Also this is not something we're planning to add in the packer as of now. We’ll keep this in mind if our goals shift!
